analyzers/src/SonarAnalyzer.CSharp/Rules/ConditionalStructureSameImplementation.cs

```diff
@@ -18,110 +18,140 @@
-                {
-                    var switchSection = (SwitchSectionSyntax)c.Node;
-
-                    if (GetStatements(switchSection).Count(IsApprovedStatement) < 2)
-                    {
-                        return;
-                    }
-
-                    var precedingSection = switchSection
-                        .GetPrecedingSections()
-                        .FirstOrDefault(preceding => CSharpEquivalenceChecker.AreEquivalent(switchSection.Statements, preceding.Statements)
-                                                     && HaveTheSameInvocations(switchSection.Statements, preceding.Statements, c.SemanticModel));
-
-                    if (precedingSection != null)
-                    {
-                        ReportSyntaxNode(c, switchSection, precedingSection, "case");
-                    }
-                },
-                SyntaxKind.SwitchSection);
-        }
+                CheckStatement(c, switchSection, precedingSections, c.SemanticModel, HasDefaultClause((SwitchStatementSyntax)switchSection.Parent), "case");
+            },
+            SyntaxKind.SwitchSection);
+    }
 
-        private static IEnumerable<StatementSyntax> GetStatements(SwitchSectionSyntax switchSection) =>
-            Enumerable.Empty<StatementSyntax>()
-                      .Union(switchSection.Statements.OfType<BlockSyntax>().SelectMany(block => block.Statements))
-                      .Union(switchSection.Statements.Where(s => !s.IsKind(SyntaxKind.Block)));
+    private static bool HasDefaultClause(SwitchStatementSyntax switchStatement) =>
+        switchStatement.Sections.Any(x => x.Labels.Any(l => l.IsKind(SyntaxKind.DefaultSwitchLabel)));
 
-        private static void CheckStatement(SonarSyntaxNodeReportingContext context, SyntaxNode statement, IEnumerable<StatementSyntax> precedingStatements)
+    private static void CheckStatement(SonarSyntaxNodeReportingContext context, SyntaxNode node, IEnumerable<SyntaxNode> precedingStatements, SemanticModel model, bool hasElse, string discriminator)
+    {
+        var numberOfStatements = GetStatementsCount(node);
+        if (!hasElse && numberOfStatements == 1)
         {
-            if (statement.ChildNodes().Count() < 2)
+            var equivalentStatements = precedingStatements.Where(x => AreEquivalentStatements(node, x, model)).ToList();
+            if (equivalentStatements.Count > 0 && equivalentStatements.Count == precedingStatements.Count())
             {
-                return;
+                ReportSyntaxNode(context, node, equivalentStatements[0], discriminator);
             }
-
-            var precedingStatement = precedingStatements.FirstOrDefault(preceding => CSharpEquivalenceChecker.AreEquivalent(statement, preceding));
-            if (precedingStatement != null)
+        }
+        else if (numberOfStatements > 1)
+        {
+            var equivalentStatement = precedingStatements.FirstOrDefault(x => AreEquivalentStatements(node, x, model));
```

Comment:
Why is this FirstOrDefault while the case before is Where/ToList()?

Reply:
I am basically bothered by my ide saying there is possible multiple enumeration before.
In the `else if` block, the IEnumerable is enumerated only once.

I will remove the `Where().ToList()` as it is a private method